File: src/lib/interpolation.ts
import type { Keyframe } from '../types';

function smoothstep(t: number): number {
  return t * t * (3 - 2 * t);
}

// Cubic bezier: x component at parameter t
function bezierX(t: number, p1x: number, p2x: number): number {
  const mt = 1 - t;
  return 3 * mt * mt * t * p1x + 3 * mt * t * t * p2x + t * t * t;
}

// Cubic bezier: y component at parameter t
function bezierY(t: number, p1y: number, p2y: number): number {
  const mt = 1 - t;
  return 3 * mt * mt * t * p1y + 3 * mt * t * t * p2y + t * t * t;
}

// Find bezier t for a given normalized x (binary search, 24 iterations ≈ 1e-7 precision)
function solveBezierT(x: number, p1x: number, p2x: number): number {
  let lo = 0, hi = 1;
  for (let i = 0; i < 24; i++) {
    const mid = (lo + hi) / 2;
    bezierX(mid, p1x, p2x) < x ? (lo = mid) : (hi = mid);
  }
  return (lo + hi) / 2;
}

export function interpolateValue(keyframes: Keyframe[], frame: number): number {
  if (keyframes.length === 0) return 0;

  const sorted = [...keyframes].sort((a, b) => a.frame - b.frame);

  if (frame <= sorted[0].frame) return sorted[0].value;
  if (frame >= sorted[sorted.length - 1].frame) return sorted[sorted.length - 1].value;

  let prev = sorted[0];
  let next = sorted[1];
  for (let i = 0; i < sorted.length - 1; i++) {
    if (frame >= sorted[i].frame && frame < sorted[i + 1].frame) {
      prev = sorted[i];
      next = sorted[i + 1];
      break;
    }
  }

  const t = (frame - prev.frame) / (next.frame - prev.frame);

  switch (prev.interpolation) {
    case 'step':
      return prev.value;
    case 'linear':
      return prev.value + (next.value - prev.value) * t;
    case 'smooth':
      return prev.value + (next.value - prev.value) * smoothstep(t);
    case 'bezier': {
      const cp1x = prev.cp1x ?? 0.25;
      const cp1y = prev.cp1y ?? 0.25;
      const cp2x = prev.cp2x ?? 0.75;
      const cp2y = prev.cp2y ?? 0.75;
      const bt = solveBezierT(t, cp1x, cp2x);
      const progress = bezierY(bt, cp1y, cp2y);
      return prev.value + (next.value - prev.value) * progress;
    }
    default:
      return prev.value + (next.value - prev.value) * t;
  }
}

File: src-tauri/src/interpolation.rs
use crate::types::{ColorKeyframe, Keyframe};

fn smoothstep(t: f64) -> f64 {
    t * t * (3.0 - 2.0 * t)
}

fn bezier_x(t: f64, p1x: f64, p2x: f64) -> f64 {
    let mt = 1.0 - t;
    3.0 * mt * mt * t * p1x + 3.0 * mt * t * t * p2x + t * t * t
}

fn bezier_y(t: f64, p1y: f64, p2y: f64) -> f64 {
    let mt = 1.0 - t;
    3.0 * mt * mt * t * p1y + 3.0 * mt * t * t * p2y + t * t * t
}

fn solve_bezier_t(x: f64, p1x: f64, p2x: f64) -> f64 {
    let (mut lo, mut hi) = (0.0_f64, 1.0_f64);
    for _ in 0..24 {
        let mid = (lo + hi) / 2.0;
        if bezier_x(mid, p1x, p2x) < x { lo = mid; } else { hi = mid; }
    }
    (lo + hi) / 2.0
}

pub fn interpolate(keyframes: &[Keyframe], frame: i64) -> f64 {
    if keyframes.is_empty() {
        return 0.0;
    }

    let mut sorted = keyframes.to_vec();
    sorted.sort_by_key(|k| k.frame);

    if frame <= sorted[0].frame {
        return sorted[0].value;
    }
    if frame >= sorted[sorted.len() - 1].frame {
        return sorted[sorted.len() - 1].value;
    }

    let mut prev = &sorted[0];
    let mut next = &sorted[1];
    for i in 0..sorted.len() - 1 {
        if frame >= sorted[i].frame && frame < sorted[i + 1].frame {
            prev = &sorted[i];
            next = &sorted[i + 1];
            break;
        }
    }

    let span = (next.frame - prev.frame) as f64;
    if span == 0.0 {
        return next.value;
    }
    let t = (frame - prev.frame) as f64 / span;

    match prev.interpolation.as_str() {
        "step" => prev.value,
        "linear" => prev.value + (next.value - prev.value) * t,
        "smooth" => prev.value + (next.value - prev.value) * smoothstep(t),
        "bezier" => {
            let p1x = prev.cp1x.unwrap_or(0.25);
            let p1y = prev.cp1y.unwrap_or(0.25);
            let p2x = prev.cp2x.unwrap_or(0.75);
            let p2y = prev.cp2y.unwrap_or(0.75);
            let bt = solve_bezier_t(t, p1x, p2x);
            let progress = bezier_y(bt, p1y, p2y);
            prev.value + (next.value - prev.value) * progress
        }
        _ => prev.value + (next.value - prev.value) * t,
    }
}

pub fn interpolate_color(kfs: &[ColorKeyframe], frame: i64) -> (f64, f64, f64, f64) {
    if kfs.is_empty() {
        return (0.0, 0.0, 0.0, 0.0);
    }
    let mut sorted = kfs.to_vec();
    sorted.sort_by_key(|k| k.frame);
    if frame <= sorted[0].frame {
        let k = &sorted[0];
        return (k.r, k.g, k.b, k.a);
    }
    let last = &sorted[sorted.len() - 1];
    if frame >= last.frame {
        return (last.r, last.g, last.b, last.a);
    }
    let mut prev = &sorted[0];
    let mut next = &sorted[1];
    for i in 0..sorted.len() - 1 {
        if frame >= sorted[i].frame && frame < sorted[i + 1].frame {
            prev = &sorted[i];
            next = &sorted[i + 1];
            break;
        }
    }
    let span = (next.frame - prev.frame) as f64;
    if span == 0.0 {
        return (next.r, next.g, next.b, next.a);
    }
    let t = (frame - prev.frame) as f64 / span;
    let tt = match prev.interpolation.as_str() {
        "step" => return (prev.r, prev.g, prev.b, prev.a),
        "smooth" => smoothstep(t),
        _ => t,
    };
    (
        prev.r + (next.r - prev.r) * tt,
        prev.g + (next.g - prev.g) * tt,
        prev.b + (next.b - prev.b) * tt,
        prev.a + (next.a - prev.a) * tt,
    )
}
